### PR TITLE
[Narwhal] add additional logs

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

1. Annotate logs with `sub_dag_index` to make debugging easier.
2. Ensure state lock is held when persisting certificate and sending it to consensus.

## Test Plan 

n/a

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/consensus/src/bullshark.rs [hunk 1]
```diff
@@ -150,44 +150,52 @@ impl ConsensusProtocol for Bullshark {
...
+        let mut total_committed_certificates = 0;
 
         // TODO: duplicated in tusk.rs
         for leader in utils::order_leaders(&self.committee, leader, state, Self::leader)
             .iter()
             .rev()
         {
-            debug!("Previous Leader {:?} has enough support", leader);
+            let sub_dag_index = state.latest_sub_dag_index + 1;
+            let _span = error_span!("bullshark_process_sub_dag", sub_dag_index);
```
Comment (akichidis) [inline narwhal/consensus/src/bullshark.rs:161]: Why an `error_span` here?

File: narwhal/primary/src/synchronizer.rs [hunk 2]
```diff
@@ -113,7 +117,12 @@ impl Inner {
         Ok(())
     }
 
-    async fn accept_certificate_internal(&self, certificate: Certificate) -> DagResult<()> {
+    // State lock must be held when calling this function.
+    async fn accept_certificate_internal(
+        &self,
+        _lock: &MutexGuard<'_, State>,
```
Comment (akichidis) [inline narwhal/primary/src/synchronizer.rs:123]: IIUC the introduced changes on the current code will improve the safety and semantics - but we don't expect to fix/reveal something from the existing codebase as we already acquire the necessary locks, right?

**Review by akichidis — APPROVED** [review]

File: narwhal/consensus/src/bullshark.rs (hunk 1 above)
Comment (mwtian) [reply narwhal/consensus/src/bullshark.rs:161]: As explained in https://github.com/MystenLabs/sui/pull/7462, having the span at `error` level ensures the span gets created and logs get annotated even if only logging at ERROR level.

File: narwhal/primary/src/synchronizer.rs (hunk 2 above)
Comment (mwtian) [reply narwhal/primary/src/synchronizer.rs:123]: Yes, this is just to help us confirm that these functions are called with locks. No locking logic is changed in this PR.

**Pull request merged**